src/instance.ts

```diff
@@ -631,55 +642,112 @@ Please use the format 'my-instance' or '${bigtable.projectName}/instances/my-ins
+   * @property {number} [options.pageSize] Maximum number of results per page.
+   * @param {string} [options.pageToken] A previously-returned page token
+   *     representing part of a larger set of results to view.
    * @param {function} callback The callback function.
    * @param {?error} callback.error An error returned while making this request.
    * @param {AppProfile[]} callback.appProfiles List of all AppProfiles.
+   * @property {object} callback.nextQuery A query object to receive more results.
    * @param {object} callback.apiResponse The full API response.
    *
    * @example <caption>include:samples/document-snippets/instance.js</caption>
    * region_tag:bigtable_get_app_profiles
    */
   getAppProfiles(
-    optionsOrCallback?: CallOptions | GetAppProfilesCallback,
+    optionsOrCallback?: GetAppProfilesOptions | GetAppProfilesCallback,
     cb?: GetAppProfilesCallback
   ): void | Promise<GetAppProfilesResponse> {
-    const gaxOptions =
-      typeof optionsOrCallback === 'object' ? optionsOrCallback : {};
+    const options =
+      typeof optionsOrCallback === 'object'
+        ? optionsOrCallback
+        : ({} as GetAppProfilesOptions);
     const callback =
       typeof optionsOrCallback === 'function' ? optionsOrCallback : cb!;
 
-    const reqOpts = {
-      parent: this.name,
-    };
+    const gaxOpts = extend(true, {}, options.gaxOptions);
+    let reqOpts = extend({}, options, {parent: this.name});
+    delete reqOpts.gaxOptions;
+
+    // Copy over pageSize and pageToken values from gaxOptions.
+    // However values set on options take precedence.
+    if (gaxOpts) {
+      reqOpts = extend(
+        {},
+        {
+          pageSize: gaxOpts.pageSize,
+          pageToken: gaxOpts.pageToken,
+        },
+        reqOpts
+      );
+      delete gaxOpts.pageSize;
+      delete gaxOpts.pageToken;
+    }
 
-    this.bigtable.request<google.bigtable.admin.v2.IAppProfile[]>(
+    this.bigtable.request<
+      google.bigtable.admin.v2.IAppProfile,
+      google.bigtable.admin.v2.IListAppProfilesResponse
+    >(
       {
         client: 'BigtableInstanceAdminClient',
         method: 'listAppProfiles',
         reqOpts,
-        gaxOpts: gaxOptions,
+        gaxOpts,
       },
-      (err, resp) => {
+      (err, IAppProfiles, nextPageRequest, apiResponse) => {
         if (err) {
           callback(err);
           return;
         }
-        const appProfiles = resp!.map(appProfileObj => {
+        const appProfiles = IAppProfiles!.map(appProfileObj => {
           const appProfile = this.appProfile(
             appProfileObj.name!.split('/').pop()!
           );
           appProfile.metadata = appProfileObj;
           return appProfile;
         });
-        callback(null, appProfiles, resp);
+        let failedLocations = null;
+        if (apiResponse) {
+          failedLocations = apiResponse.failedLocations!;
+        }
+        callback(
```

Comment:
This is tricky, since we let `failedLocations` be on its own in the last PR, as well as BigQuery's `insertErrors`, I believe, which is a similar concept. However, we are also breaking consistency from all of our other paginated methods, which use `err, objects, nextQuery, apiResponse`.  I think we should find a way to get this down to the usual callback length by either combining `appProfiles` and `failedLocations`, or just documenting where the user can find the failed locations in the `apiResponse`. What do you think?

Reply:
The inconsistency comes from the fact that `next_page_token` is deprecated from listClusters/InstancesResponse, which makes `nextQuery` obsolete for those functions. However `next_page_token` is supported in all other paginated responses.

I think, for consistency purposes, we can add `nextQuery` property to getClusters/Instances response signature (it will alway be empty) and bring the response to a consistent across the repo form of `err, objects, failedLocations, nextQuery, apiResponse`.
Otherwise documenting where failed locations could be found consistency wise I think is a better option. What do you think?

Reply:
> The inconsistency comes from the fact that next_page_token is deprecated from listClusters/InstancesResponse, which makes nextQuery obsolete for those functions.

Yeah, and that's fine, so nothing we can do there. Having a new callback parameter "failedLocations" was a good idea 👍 

But just so I understand here, this method *does* have pagination, right? "nextPageRequest" is basically "nextQuery"?

> Otherwise documenting where failed locations could be found consistency wise I think is a better option.

I would lean towards the documentation option. I think it would mislead users to return the defunct `nextQuery`.

Reply:
>"nextPageRequest" is basically "nextQuery"?

`nextPageRequest` is the `IListObjectsRequest` in this case `IListAppProfilesRequest`.
The `nextPageRequest` is only returned when there are more results available.
 
>I would lean towards the documentation option. I think it would mislead users to return the defunct nextQuery.

Empty object for nextQuery/nextPageRequest also signals that there are no more results to be returned. I won't disrupt the workflow
something like
```ts
      const gaxOptions = {autoPaginate: false};
      let allObjects: AppProfile[] = [];
      let getObjectsOptions = {
        pageSize: 1,
        pageToken: '',
        gaxOptions,
      };
      while (true) {
        const [objects, , next] = await instance.getObjects(
          getObjectsOptions
        );
        allObjects = [...allObjects, ...objects];
        if (!next) {
          break;
        }
        getObjectsOptions = Object.assign(
          {},
          {gaxOptions},
          {pageSize: next.pageSize!, pageToken: next.pageToken!}
        );
      }
```
WDYT?

Reply:
> The nextPageRequest is only returned when there are more results available.

Will the user just provide that argument directly, like they do our nextQuery object?:

```js
getAppProfiles(config, function(err, appProfiles, nextQuery) {
  if (nextQuery) {
    getAppProfile(nextQuery, function(err, nextAppProfiles, nextNextQuery) {})
  }
})
```

Our `nextQuery` object should always be an extension of the user-provided `c onfig` with the appropriate settings for cursoring. We should stick to that system, if we're not already, by passing the `IListAppProfilesRequest` back.

> Empty object for nextQuery/nextPageRequest also signals that there are no more results to be returned.

That's true, but it's misleading to make users write code anticipating more results could exist, when we know they wouldn't ever. That's also a little confusing to maintain down the road. The main thing I'm avoiding is the high number of response arguments. We've usually tried to consolidate the response pattern:

- error
- the most logical thing the user expects
- the full API response for whenever they need something else

For query methods, we added in "nextQuery" to make cursoring easier:

- error
- the most logical thing the user expects
- an object ready to pass to the API to continue getting results
- the full API response for whenever they need something else

So, I'm hesitant to squeeze one more in there, as if it's not "the most logical thing the user expects", then we've historically pointed them to "the full API response for whenever they need something else".

One more thought-- if the user runs this with autoPagination: true (which is on by default), doesn't that abolish all `apiResponse`s? In this sample:

```js
getAppProfiles()
  .then(result => {
    const appProfiles = result[0];
    if (result[1]) {
      // These locations contain instances which could not be retrieved.
      const failedLocations = result[1];
    }
  })
```

What actually is `failedLocations`? An array of arrays of strings?

My vote would go to documenting how to find `failedLocations` in the API response, but more input on this one could be useful.

Reply:
> Our nextQuery object should always be an extension of the user-provided `config` with the appropriate settings for cursoring. We should stick to that system, if we're not already, by passing the IListAppProfilesRequest back.

Combined `IListAppProfilesRequest` with user-provided `config`, now below sample will work.
```ts
getAppProfiles(config, function(err, appProfiles, nextQuery) {
  if (nextQuery) {
    getAppProfile(nextQuery, function(err, nextAppProfiles, nextNextQuery) {})
  }
})
```

> My vote would go to documenting how to find failedLocations in the API response, but more input on this one could be useful.

Removed `failedLocations` from the response and documented in the callback section instead.

PTAL

Reply:
> How shall the callback look `{err, resources, failedLocations, fullResponse}` or just `{err, resources, fullResponse}` and document where to find `failedLocations` if any?

Very good question. I think there's a good case for both options. Have we released yet with those changes, where if we change the response signature, we'd be breaking? If we have to break, I'd probably just leave those alone. If not, I would choose `(err, resources, apiResponse)`:

- 👍 Simple signature
- 👍 Consistent with most of our methods, where we return "error, thing you want, everything you could ever want"
- 👎 Have to find `failedLocations` inside of `apiResponse`

Reply:
Great, will revert to same in `bigtable.getInstances()`